File: Arashyn-FE/src/shared/responses/grammar_detail_response.ts
export type Example = {
    id: string;
    sentence: string;
    translation: string;
    note?: string;
};

export type Meaning = {
    id: string;
    content: string;
    groupKey: number;
    examples: Example[];
};

export type Component = {
    id: string;
    order: number;
    keyword: string | null;
    form: string | null;
    groupKey: number;
    optional: boolean;
};

export type ComponentGroup = {
    groupKey: number;
    components: Component[];
    meanings: Meaning[];
};

export type Note = {
    id: string;
    content: string;
};

export type Filter = {
    id: string;
    name: string;
}

export type grammar_detail_response = {
    id: string;
    title: string;
    language: string;
    isPublic: boolean;
    ownerId: string;
    ownerName: string;
    groups: ComponentGroup[];
    notes: Note[];
    filters: Filter[];
};

File: Arashyn-BE/src/main/java/com/arashi/edu/arashynbe/features/system/grammar/dto/response/GrammarDetailResponse.java
package com.arashi.edu.arashynbe.features.system.grammar.dto.response;

import com.arashi.edu.arashynbe.features.system.component.dto.response.GrammarComponentResponse;
import com.arashi.edu.arashynbe.features.system.filter.dto.response.GrammarFilterResponse;
import com.arashi.edu.arashynbe.features.system.meaning.dto.response.GrammarMeaningResponse;
import com.arashi.edu.arashynbe.features.system.note.dto.response.GrammarNoteResponse;
import com.arashi.edu.arashynbe.shared.enums.Language;

import java.util.List;
import java.util.UUID;

public record GrammarDetailResponse(

        UUID id,

        String title,

        Language language,

        boolean isPublic,

        UUID ownerId,

        String ownerName,

        List<Group> groups,

        List<GrammarNoteResponse> notes,

        List<GrammarFilterResponse> filters

) {

  public record Group(

          Integer groupKey,

          List<GrammarComponentResponse> components,

          List<GrammarMeaningResponse> meanings

  ) {
  }
}